### ISSUE TITLE
Broken? ccache on travis

### ISSUE THREAD

**Issue by gdevenyi** [packpack/packpack#13]
Finding these in my build outputs:
Ubuntu trusty:
```
ccache: failed to create /root/.ccache (Permission denied)
```
Centos
```
ccache: failed to create /cache/ccache (No such file or directory)
```

Fedora:
```
ccache: error: Failed to create directory /cache/ccache/tmp: Permission denied
```

**Comment by gdevenyi** [comment]
Probably causing:
```
configure: error: C compiler cannot create executables
```

In my builds

**Comment by gdevenyi** [comment]
Possibly related to #9 ?

**Comment by benoit-pierre** [comment]
This is the problem:
```diff
diff --git i/packpack w/packpack
index 21bf514..a43ee53 100755
--- i/packpack
+++ w/packpack
@@ -122,7 +122,7 @@ docker run \
         --entrypoint=/build/userwrapper.sh \
         -e XDG_CACHE_HOME=/cache \
         -e CCACHE_DIR=/cache/ccache \
-        --volume "${HOME}/.cache:/ccache" \
+        --volume "${HOME}/.cache:/cache" \
         ${DOCKER_REPO}:${DOCKER_IMAGE} \
         make -f /pack/Makefile -C /source BUILDDIR=/build -j $@
```
Tested with `OS=ubuntu DIST=trusty ./packpack -f test.mk test`, with this in `test.mk`:
```make
test:
        ccache gcc -x c /dev/null -c -o /build/test.o
```

**Comment by benoit-pierre** [comment]
Also, might be better to just reuse the `CACHE_DIR` variable:
```diff
diff --git i/packpack w/packpack
index 21bf514..e648eda 100755
--- i/packpack
+++ w/packpack
@@ -122,7 +122,7 @@ docker run \
         --entrypoint=/build/userwrapper.sh \
         -e XDG_CACHE_HOME=/cache \
         -e CCACHE_DIR=/cache/ccache \
-        --volume "${HOME}/.cache:/ccache" \
+        --volume "${CACHE_DIR}:/cache" \
         ${DOCKER_REPO}:${DOCKER_IMAGE} \
         make -f /pack/Makefile -C /source BUILDDIR=/build -j $@
```

**Comment by benoit-pierre** [comment]
@gdevenyi: the fact that in your build with Ubuntu Trusty ccache tries to use `/root/.ccache` is strange, is `CCACHE_DIR` overridden somewhere in your environment?

**Comment by gdevenyi** [comment]
Nope, everything is normal as far as I can tell:
https://github.com/gdevenyi/minc-toolkit-v2/blob/auto-build/.travis.yml

**Comment by gdevenyi** [comment]
Bigger problem: due to the way we're using travis.yml, it looks like the cache will never be used since it's synced *before* the deploy step, which is actually where we build!:

```
The command "git submodule update --init --recursive" exited with 0.
store build cache

change detected (content changed, file is created, or file is deleted):
/home/travis/.ccache/c/stats


changes detected, packing new archive
uploading archive

$ git clone https://github.com/packpack/packpack.git packpack
Cloning into 'packpack'...
remote: Counting objects: 1018, done.
remote: Compressing objects: 100% (9/9), done.
remote: Total 1018 (delta 2), reused 0 (delta 0), pack-reused 1009
Receiving objects: 100% (1018/1018), 922.31 KiB | 0 bytes/s, done.
Resolving deltas: 100% (566/566), done.
Checking connectivity... done.

$ packpack/packpack
+docker run --volume /home/travis/build/gdevenyi/minc-toolkit-v2/packpack/pack:/pack:ro --volume /home/travis/build/gdevenyi/minc-toolkit-v2:/source:ro --volume /home/travis/build/gdevenyi/minc-toolkit-v2/build:/build --env-file /home/travis/build/gdevenyi/minc-toolkit-v2/build/env --workdir /source --rm=true --entrypoint=/build/userwrapper.sh -e XDG_CACHE_HOME=/cache -e CCACHE_DIR=/cache/ccache --volume /home/travis/.cache:/ccache packpack/packpack:ubuntu-trusty make -f /pack/Makefile -C /source BUILDDIR=/build -j
Unable to find image 'packpack/packpack:ubuntu-trusty' locally
ubuntu-trusty: Pulling from packpack/packpack

```

**Comment by gdevenyi** [comment]
We're also exporting the CCACHE_DIR wrong into docker:
```
-e CCACHE_DIR=/cache/ccache
```

Should be
```
-e CCACHE_DIR=/cache
```

**Comment by benoit-pierre** [comment]
No, "/cache" is `XDG_CACHE_HOME` (`~/.cache/packpack`).

**Comment by gdevenyi** [comment]
^^^See edit

**Comment by benoit-pierre** [comment]
See my patch :P. The goal is to store all caches in `XDG_CACHE_HOME/packpack` (with ccache cache in `XDG_CACHE_HOME/packpack/ccache`, with `XDG_CACHE_HOME` set to `/cache` in docker so tools that honor `XDG_CACHE_HOME` store their cache in a persistent location.

**Comment by gdevenyi** [comment]
Okay, I can't find any docs that say travis will honour XDG_CACHE_HOME.

With the current travis config, it appears as though /home/travis/.ccache is cached.

The other challenge of course is that right now, it appears that the cache is never populated because it's packed up after "script" and we don't start doing work until deploy.
See https://docs.travis-ci.com/user/caching/#Build-phases

**Comment by benoit-pierre** [comment]
The patch has nothing to do with Travis, it's so other caches, like pip, are persistent, when you run packpack multiple times. For caching in Travis, AFAIK:
* you can specify what directories to cache
* you can only cache results of the build phase, not the deploy phase

**Comment by gdevenyi** [comment]
Okay.

I get what you're trying to do.

Things we need to handle:
- Travis properly caching
- Honouring local cache directories if they exist

**Comment by benoit-pierre** [comment]
I think something like [this](https://github.com/benoit-pierre/minc-toolkit-v2/blob/auto-build/.travis.yml) might work, except maybe a different `CACHE_NAME` needs to be set for each environment (like [here](https://github.com/openstenoproject/plover/blob/master/.travis.yml)).

**Comment by rtsisyk** [comment]
Fixed by 6493310e7863494d095b4a877e7dcc2c3ca52ad0 and f7ae60a438f3b6e8bf05dcdc61aea444b23d2ed3
